### PR TITLE
enh(api) : add the posibility to search using host categories & host & host groups

### PR DESCRIPTION
**Pull request by Exodiaaaa**
## Description

Enhance endpoint GET: /configuration/services/groups by adding possibility to search by host categories

**Fixes** # (MON-34391)

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [X] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 22.04.x
- [ ] 22.10.x
- [ ] 23.04.x
- [ ] 23.10.x
- [X] 24.04.x (master)

<h2> How this pull request can be tested ? </h2>

try to filter hostgroups using the new option on the filter : http://localhost:8080/centreon/api/v24.04/configuration/services/groups?search={"host_id":1} /{"host_group_id":1}/{"host_category_id":1}
You should create a hostcategorie to test it

#### Community contributors & Centreon team

- [X] I have followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [X] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php [hunk 1]
```diff
@@ -323,12 +323,50 @@ private function getFindServiceGroupConcatenator(array $accessGroupIds = []): Sq
...
+                            ON arsr.acl_res_id = res.acl_res_id
+                    SQL
+            )
             ->defineOrderBy(
                 <<<'SQL'
                     ORDER BY sg.sg_name ASC
                     SQL
             );
 
+        if (! empty($_GET['search']) && mb_strpos($_GET['search'], 'host_group_id')) {
```
Comment (adr-mo) [inline centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php:340]: @Exodiaaaa Hi,
⚠️ For repositories to handle the search request parameter we do not use the global GET variable ⚠️ 
Please see how it is done in this example https://github.com/centreon/centreon/blob/2d3e27f9e7c06cf57af268cbb7f82dc74f9c02bc/centreon/src/Core/Host/Infrastructure/Repository/DbReadHostRepository.php#L446

**Review by adr-mo — CHANGES_REQUESTED** [review]

**Review by paul-oureib — DISMISSED** [review]

File: centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php [hunk 2]
```diff
@@ -390,7 +390,51 @@ private function retrieveServiceGroups(
...
+                    LEFT JOIN `:db`.acl_resources_sg_relations arsr
+                        ON sg.sg_id = arsr.sg_id
+                    LEFT JOIN `:db`.acl_resources res
+                            ON arsr.acl_res_id = res.acl_res_id
+                    LEFT JOIN `:db`.acl_resources_hc_relations arhcr
+                        ON res.acl_res_id = arhcr.acl_res_id
+                    LEFT JOIN `:db`.hostcategories hc
+                        ON arhcr.hc_id = hc.hc_id
+                        AND hc.level IS NOT NULL
+                    SQL
```
Comment (a-launois) [inline centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php:435]: ⚠️ This doesn't work
The servicegroups listed will be in the same ACLs Resource group as the host_group/host/host_category but it does'nt mean they are linked in terms of resource configuration
You need to join on the relation tables of servicegroup-service, service-host, host-hostgroup (for example)

But as in you other PRs on the same subject:
⚠️ If you are in an on-prem platform with non-admin user,
you have to filter hostgroups, host and hostcategories on the user's ACLs

I talked about it with @dmyios, you can see with him for more detailed explanations

**Review by a-launois — CHANGES_REQUESTED** [review]

File: centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php [hunk 3]
```diff
@@ -66,6 +66,32 @@ public function findAll(?RequestParametersInterface $requestParameters): \Traver
...
+            )
+            ->appendJoins(
+                <<<'SQL'
+                        LEFT JOIN `:db`.hostgroup_relation hgr
+                            ON  h.host_id = hgr.host_host_id
+                        LEFT JOIN `:db`.hostgroup hg
+                            ON hgr.hostgroup_hg_id = hg.hg_id
+                    SQL
+            );
```
Comment (jeremyjaouen) [inline centreon/src/Core/ServiceGroup/Infrastructure/Repository/DbReadServiceGroupRepository.php:93]: Like for the other PR, check that the filter is present before doing the Join

**Review by a-launois — CHANGES_REQUESTED** [review]
I suggest that you check what Dmytro did on this two PRs:
https://github.com/centreon/centreon/pull/3351
https://github.com/centreon/centreon/pull/3407
The main attention points being the ACLs 
- on host (check if user has access to all hosts check) 
- as well as groups/categories (check if user has filtered access)

**Comment by adr-mo** [comment]
Closing this PR as I take back the subject. Thanks @Exodiaaaa for the help 💯

**Pull request closed without merging**